File: src/discord/commands/public/moderate/utils/delete-message.ts
import { res } from "#functions";
import { deleteMessageMenu, loadMenu, closeMenu } from "#menus";
import { createEmbed } from "@magicyan/discord";
import { User, GuildTextBasedChannel, Guild, GuildMember, InteractionReplyOptions, ChatInputCommandInteraction, ComponentType } from "discord.js";
import { filterUsers } from "./index.js";


async function deleteMessageAction<R>(user: User, ids: string[], channel: GuildTextBasedChannel, guild: Guild): Promise<R> {
    let message_owners: GuildMember[] = [];
    let failed: string[] = [];

    let messages = await channel.messages.fetch({ limit: 100 });
    for (const id of ids) {
        const member = guild.members.cache.get(id);
        if (!member) {
            failed.push(id);
            continue;
        }
        message_owners.push(member);
    }

    if (message_owners.length == 0) {
        const embed = createEmbed({
            color: "Blue",
            author: {
                name: user.globalName || user.username,
                iconURL: user.avatarURL() || undefined
            },
            title: "**Officer Cui's panel**",
            thumbnail: "https://raw.githubusercontent.com/Cesio137/Bangboo/refs/heads/master/assets/avatar/Officer.png",
            description: "**Failed to delete message(s): No user(s) selected**"
        });

        return ({
            flags: ["Ephemeral"],
            embeds: [embed],
            components: [],
        } satisfies InteractionReplyOptions) as R;
    }

    messages = messages.filter(msg => msg.member && message_owners.indexOf(msg.member) != -1);
    if (messages.size > 0) await channel.bulkDelete(messages, true);

    let description = `**Deleted message(s) from users:**\n${message_owners.map(id => `<${id}>`).join("\n")}`;
    if (failed.length > 0) {
        description = `${description}\n`
        description = `**Failed to delete message(s) from user(s):**\n${failed.map(id => `<${id}>`).join("\n")}`
    }

    const embed = createEmbed({
        color: "Blue",
        author: {
            name: user.globalName || user.username,
            iconURL: user.avatarURL() || undefined
        },
        title: "**Officer Cui's panel**",
        thumbnail: "https://raw.githubusercontent.com/Cesio137/Bangboo/refs/heads/master/assets/avatar/Officer.png",
        description: description
    });

    return ({
        flags: ["Ephemeral"],
        embeds: [embed],
        components: [],
    } satisfies InteractionReplyOptions) as R;
}

export async function deleteMessageCollector(interaction: ChatInputCommandInteraction<"cached">) {
    const { channel, guild, user } = interaction;
    if (!channel) {
        interaction.reply(
            res.danger("Invalid channel.")
        )
        return 
    }

    let ids: string[] = [];
    let timeout = true;
    let isOk = false;

    await interaction.deferReply({ flags: ["Ephemeral"] });
    let messageID = (await interaction.fetchReply()).id;
    interaction.editReply(deleteMessageMenu(user, ids));

    const time = Date.now() + 300000;

    const userCollector = channel.createMessageComponentCollector(
        {
            componentType: ComponentType.UserSelect,
            filter: (componentInteraction) => componentInteraction.user.id === interaction.user.id && componentInteraction.message.id === messageID,
            time: time - Date.now(),
        }
    );

    const btnCollector = channel.createMessageComponentCollector(
        {
            componentType: ComponentType.Button,
            filter: (componentInteraction) => componentInteraction.user.id === interaction.user.id && componentInteraction.message.id === messageID,
            time: time - Date.now(),
        }
    );
    
    userCollector.on("collect", async function(i) {
        const { customId, guild, user } = i;
        if (!guild) { return }
        switch(customId) {
            case "mod/select-users":
                await i.update(loadMenu(user, "👥 **Filtering selected users...**"));
                ids = filterUsers(i.values, guild);
                await i.editReply(deleteMessageMenu(user, ids));
                break;
        }
    });
    
    btnCollector.on("collect", async function(i) {
        const { customId, user } = i;
        switch(customId) {
            case "mod/btn-cancel":
                timeout = false;
                userCollector.stop(); btnCollector.stop();
                break;

            case "mod/btn-confirm":
                await i.update(loadMenu(user, "👥 **Deleting messages from selected users...**"))
                await i.editReply(await deleteMessageAction(user, ids, channel, guild));
                timeout = false;
                isOk = true;
                userCollector.stop(); btnCollector.stop();
                break;
        }
    });

    btnCollector.on("end", async function() {
        if (!isOk) interaction.editReply(closeMenu(interaction.user, timeout));
    });
}

File: src/discord/commands/public/moderate/delete_message.rs
use crate::discord::*;
use crate::helpers::*;
use crate::menus::*;
use crate::utils::*;
use serenity::all::{
    CacheHttp, CommandInteraction, ComponentInteraction, ComponentInteractionCollector,
    ComponentInteractionDataKind, Context, CreateEmbed, CreateEmbedAuthor, Member, MessageFlags,
    UserId,
};
use serenity::futures::StreamExt;
use serenity::nonmax::NonMaxU8;
use std::time::Duration;

pub async fn delete_message_action(
    ctx: &Context,
    interaction: &ComponentInteraction,
    ids: &[UserId],
) {
    let mut payload = ReplyPayload::default();
    payload.components = Some(vec![]);

    let mut success: Vec<UserId> = Vec::new();
    let mut failed: Vec<UserId> = Vec::new();

    let user = &interaction.user;

    let max_msg = NonMaxU8::new(255);
    let message_ids = match ctx
        .http
        .get_messages(interaction.channel_id, None, max_msg)
        .await
    {
        Ok(msg) => msg
            .iter()
            .filter(|val| ids.contains(&val.author.id))
            .cloned()
            .map(|val| val.id)
            .collect::<Vec<_>>(),
        Err(_) => {
            let embed_author =
                CreateEmbedAuthor::new(user.global_name.as_ref().unwrap_or(&user.name))
                    .icon_url(user.avatar_url().unwrap_or(user.default_avatar_url()));

            let embed = CreateEmbed::new()
                .color(str_hex_to_u32(&CONSTANTS.colors.royal))
                .author(embed_author)
                .title("**Officer Cui's panel**")
                .thumbnail("https://raw.githubusercontent.com/Cesio137/Bangboo/refs/heads/rust/assets/avatar/Officer.png")
                .description("**Failed to delete message(s): No user(s) selected**");

            payload.embeds = Some(vec![embed]);
            update_component(ctx, interaction, MessageFlags::EPHEMERAL, &payload).await;
            return;
        }
    };

    match interaction
        .channel_id
        .delete_messages(ctx.http(), &message_ids, None)
        .await
    {
        Ok(_) => success.extend(ids),
        Err(err) => {
            error(&format!("Could not delete message(s)\n└ {:?}", err));
            failed.extend(ids)
        }
    }

    let mut description = Vec::new();
    if !success.is_empty() {
        description.push("**Deleted message(s) from users:**".to_string());
        for user_id in success {
            description.push(format!("<@{}>", user_id));
        }
    }
    if !failed.is_empty() {
        description.push("\n**Failed to delete message(s) from user(s):**".to_string());
        for user_id in failed {
            description.push(format!("<@{}>", user_id));
        }
    }
    let description = description.join("\n");

    let embed_author = CreateEmbedAuthor::new(user.global_name.as_ref().unwrap_or(&user.name))
        .icon_url(user.avatar_url().unwrap_or(user.default_avatar_url()));

    let embed = CreateEmbed::new()
        .color(str_hex_to_u32(&CONSTANTS.colors.royal))
        .author(embed_author)
        .title("**Officer Cui's panel**")
        .thumbnail("https://raw.githubusercontent.com/Cesio137/Bangboo/refs/heads/rust/assets/avatar/Officer.png")
        .description(description);
    payload.embeds = Some(vec![embed]);
    update_component(ctx, &interaction, MessageFlags::EPHEMERAL, &payload).await;
}

pub async fn delete_message_collector(
    ctx: &Context,
    interaction: &CommandInteraction,
    member: &Box<Member>,
) {
    let mut payload = ReplyPayload::default();
    let empty_slice: Vec<UserId> = Vec::new();
    let (embed, components) = message_delete_menu(&member.user, &empty_slice);
    payload.embeds = Some(vec![embed]);
    payload.components = Some(components);
    if !reply(ctx, interaction, MessageFlags::EPHEMERAL, &payload).await {
        return;
    }

    let message_id = match interaction.get_response(ctx.http()).await {
        Ok(msg) => msg.id,
        Err(_) => {
            error("Failed to get message id.");
            return;
        }
    };
    let user_id = member.user.id;
    let filter = move |i: &ComponentInteraction| {
        i.message.id == message_id && i.member.as_ref().unwrap().user.id == user_id
    };

    // Data
    let mut ids: Vec<UserId> = Vec::new();
    let mut cancel = false;
    let mut timeout = true;

    let mut collector = ComponentInteractionCollector::new(&ctx)
        .filter(filter)
        .author_id(interaction.user.id)
        .timeout(Duration::from_secs(300))
        .stream();

    while let Some(i) = collector.next().await {
        payload = ReplyPayload::default();

        match &i.data.kind {
            ComponentInteractionDataKind::Button => {
                if &i.data.custom_id == "mod/btn-cancel" {
                    timeout = false;
                    cancel = true;
                    break;
                }
                _ = i.defer_ephemeral(ctx.http());
                delete_message_action(ctx, &i, &ids).await;
                timeout = false;
                break;
            }
            ComponentInteractionDataKind::UserSelect { values } => {
                ids = values.to_vec();
            }
            _ => {}
        }

        let (embed, components) = message_delete_menu(&member.user, &ids);
        payload.embeds = Some(vec![embed]);
        payload.components = Some(components);
        update_component(ctx, &i, MessageFlags::EPHEMERAL, &payload).await;
    }
    if timeout || cancel {
        payload = ReplyPayload::default();
        let close_embed = close_menu(&member.user, timeout);
        payload.embeds = Some(vec![close_embed]);
        payload.components = Some(vec![]);
        edit(ctx, interaction, MessageFlags::EPHEMERAL, &payload).await;
    }
}
